pnpm.fetchDeps: init

Scrumplex opened the pull request:
## Description of changes


Nowadays [pnpm](https://pnpm.io/) is almost unavoidable when working with any larger Node.js project. In an effort to support this package manager in Nixpkgs, this PR introduces `pnpm.fetchDeps` which produces a FOD containing a pnpm store that can be used for a reproducible offline build.

Closes #231513

### To-dos
- [ ] Disable `withNode` for `pnpm` when used in a derivation
- [x] Add a `pnpm server start` script to `pnpmDeps` FODs
- [ ] Support `pnpm-workspace.yaml` projects

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Scrumplex committed:
pnpm.fetchDeps: init

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

getchoo commented:
this could be a near drop-in replacement for methods used in [geph](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/networking/geph/default.nix#L63), [kiwitalk](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/ki/kiwitalk/package.nix#L37), [pot](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/misc/pot/default.nix#L43), [vesktop](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/ve/vesktop/package.nix#L38), and [youtube-music](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/audio/youtube-music/default.nix#L32)

dotlambda commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,87 @@
+
+        export HOME=$(mktemp -d)
+        pnpm config set store-dir $out
+        pnpm install --frozen-lockfile --ignore-script
```
What guarantee do we have that this is reproducible across versions of pnpm?

Scrumplex replied:
I don't think we have any guarantees here. So far `vesktop` has been using this, and it seems to be fairly stable

wegank replied:
I don't think this should prevent the PR from merging. There is no guarantee between versions of Maven either, but we can always update the hashes in `buildMavenPackage` treewide in the PR that updates `maven`.

Scrumplex replied:
I have now pinned the pnpm version used by this function. This will almost certainly guarantee reasonable<sup>1</sup> reproducibility.

<sup>1</sup>: Perhaps different Node.js versions could influence the reproducibility, but I would argue that it's reasonably reproducible nevertheless.

Scrumplex commented:
I got rid of the `supportedArchitectures` code, as `--force` achieves the same without having to specify all platforms

Septias commented:
anything stopping this from merging?

dotlambda commented:
> anything stopping this from merging?

Yes. You should read.

Septias commented:
Ah, I thought that issue was resolved because of the follow-up commits and `outdated`. But seems like it still exists. Sorry for the confusion.

rafaelsgirao commented on pkgs/build-support/node/fetch-pnpm-deps/pnpm-config-hook.sh:
```diff
@@ -0,0 +1,36 @@
+    echo "Executing pnpmConfigHook"
+
+    if [ -z "${pnpmDeps-}" ]; then
+        echo "Error: 'npmDeps' should be set when using npmConfigHook."
```
```suggestion
        echo "Error: 'pnpmDeps' should be set when using pnpmConfigHook."
```

nyabinary commented:
Needs a rebased :3

Scrumplex commented:
I think there would be two ways to keep this FOD deterministic:

1. Pin pnpm to some version
2. Write a program to fetch pnpm deps, like `fetchNpmDeps` does

For 1., We should probably version the function, so `fetchPnpmDeps` could become `fetchPnpmDeps9` (assuming we pin it to a version that supports lock file version 9). Consumers only need to "update" to the latest version of this function if they need support for a newer lockfile version. I could see this break down if we need to bump the pinned version because of a security vulnerability or if there are minor lock file version bumps (i.e. 9.0 -> 9.1)

pluiedev commented:
I definitely feel like using a versioned function makes way more sense, given how often pnpm updates its lockfile format — plus, not all projects immediately update to the latest lockfile version anyway :) We can just remove old versions if all pnpm versions that use it go EOL so having too many (largely) duplicate functions shouldn't be *that* bad

szlend commented:
I agree that versioned fetched are the way to go with how often things change in the javascript package management world. There's just no way we can keep up with maintaining custom lock parsers and fetchers. Same issue with Yarn in https://github.com/NixOS/nixpkgs/issues/254369

doronbehar commented:
I agree with the above comments, and I think we can move forward with this PR in it's current form. It'd be nice to add documentation to `doc/languages-frameworks/javascript.section.md` about this function, with an example or two, and add a release-notes entry.

> this could be a near drop-in replacement for methods used in [geph](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/networking/geph/default.nix?rgh-link-date=2024-02-22T22%3A47%3A46Z#L63), [kiwitalk](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/ki/kiwitalk/package.nix?rgh-link-date=2024-02-22T22%3A47%3A46Z#L37), [pot](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/misc/pot/default.nix?rgh-link-date=2024-02-22T22%3A47%3A46Z#L43), [vesktop](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/ve/vesktop/package.nix?rgh-link-date=2024-02-22T22%3A47%3A46Z#L38), and [youtube-music](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/audio/youtube-music/default.nix?rgh-link-date=2024-02-22T22%3A47%3A46Z#L32)
> 
> edit: can also be used in [gitbutler](https://github.com/NixOS/nixpkgs/pull/289664) and [modrinth-app](https://github.com/NixOS/nixpkgs/pull/289149)

It'd be nice to practice this new `fetchPnpmDeps` function on the packages mentioned here ^ - this would also give us confidence that the FODs it produces are solid enough. I personally would like to package [`bash-languague-server`](https://github.com/bash-lsp/bash-language-server) with such a `fetchPnpmDeps`, and it'd also be nice to see how is that packaged.

We could also wait a few weeks after you @Scrumplex have built these FODs, and I will try to build them as well afterwards with `nix build --rebuild $attr.pnpmDeps` to test the FOD hashes.

Scrumplex committed:
nodePackages: enable overriding src and name attrs

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
pot: use pnpm.fetchDeps

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
kiwitalk: use pnpm.fetchDeps

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
gephgui-wry: use pnpm.fetchDeps

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
youtube-music: use pnpm.fetchDeps

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

doronbehar commented on pkgs/applications/audio/youtube-music/default.nix:
```diff
@@ -24,54 +21,17 @@ stdenv.mkDerivation (finalAttrs: {
-      aarch64-linux = "sha256-0kyjjttpXpFVhdza5NAjGrRn++qc/N5/u2dQl7VufLE=";
-      x86_64-darwin = "sha256-Q37QJt/mhfpSguOlkJGKFTCrIOrpbG3OBwaD/Bg09Us=";
-      aarch64-darwin = "sha256-wbfjzoGa/6vIlOOVX3bKNQ2uxzph3WSofo3MGXqA6yQ=";
-    }.${stdenv.system} or (throw "Unsupported system: ${stdenv.system}");
```
We should check that there is no hash mismatch for different systems, now when `fetchPnpmDeps` is used.

doronbehar replied:
Ofbrorg is reporting the hashes mismatch:

On aarch64-linux:

```
error: hash mismatch in fixed-output derivation '/nix/store/qq4ncfffg02xdx2laqcx1fclk0vjqxfn-youtube-music-pnpm-6_0-deps.drv':
         specified: sha256-EMIzOzX1QDNDcGRTV189SXiH+lZp0PgAFX5IERBuilI=
            got:    sha256-egfteCl9ThMNz/mlYVFQTfwmIwHIbP1/0V2dhV8BL0E=
```

Later ofborg will report the hash mismatches from other platforms.

Scrumplex replied:
I see. Interestingly enough this does not affect all packages here. Vesktop for example works fine when built on Linux aarch64 and x86_64.

The issue seems to be a "side effects cache" where some packages produce platform-dependent outputs. It seems like I can disable that though, which might work here

Scrumplex replied:
Could both of you try again?

szlend replied:
1 package built:
youtube-music

Result of `nixpkgs-review pr 290715` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>youtube-music</li>
  </ul>
</details>

doronbehar replied:
> Could both of you try again?

CI seems totally green for `{aarch64,x86_64}-linux`, but there is a wierd error on `x86_64-darwin`.

wegank replied:
That error is #294725, we can ignore it for now.

doronbehar commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,101 @@
+
+    __functor = self: self.fetchPnpmDeps;
+  };
+in {
+  fetchPnpmDeps_6_0 = mkFetchPnpmDeps "6_0";
+  fetchPnpmDeps_7_0 = mkFetchPnpmDeps "7_0";
+}
```
I think you could also create this attribute set from the attribute keys of `pinnedVersions`.

Scrumplex replied:
Yes, I was actually torn on this because I feel like it is more readable this way, but I guess the following Nix expression is also straighforward:
```nix
mapAttrs' (version: _: nameValuePair "fetchPnpmDeps_${version}" (mkFetchPnpmDeps version))
```

doronbehar replied:
> Yes, I was actually torn on this because I feel like it is more readable this way, but I guess the following Nix expression is also straighforward:
> 
> ```nix
> mapAttrs' (version: _: nameValuePair "fetchPnpmDeps_${version}" (mkFetchPnpmDeps version)) pnpms
> ```

Very nit :). I personally prefer this, as it is generalizing. You also inherit `fetchPnpmDeps_..` attributes in `all-packages.nix`, so it helps a bit to understand. You could also give a comment near this line.

doronbehar commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,101 @@
+, makeSetupHook
+}: let
+  # The names of these attributes refer to the pnpm lockfile version
+  pinnedVersions = {
```
Maybe this attrset could better reflect what it is about.. How about `lockfileVerPnpm`? or `lockVerPnpm`?

getchoo replied:
i would say this name is fitting. after all, the contents of the attribute set *are* the pinned versions. the only issue here is the naming of the attributes, but that's explained in the above comment and i think would be made apparent when used in conjunction with functions that have the versions in their name

doronbehar commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,101 @@
+    };
+  };
+
+  pnpms = builtins.mapAttrs (_: value: nodePackages.pnpm.overrideAttrs value) pinnedVersions;
```
Are these attributes useable for something other then `fetchPnpmDeps`? Perhaps they should be accessible in `all-packages` as well?

Scrumplex replied:
They are *technically* accessible already using `fetchPnpmDeps_${version}.pnpm`. I don't think `nodePackages` supports versioned packages like this, though, correct me if I am wrong.

doronbehar reviewed (commented):
Looks great @Scrumplex ! These are my only comments / questions / doubts.

Scrumplex commented on pkgs/top-level/all-packages.nix:
```diff
@@ -10138,6 +10138,9 @@ with pkgs;
   importNpmLock = callPackages ../build-support/node/import-npm-lock { };
 
+  inherit (callPackages ../build-support/node/fetch-pnpm-deps { })
+    fetchPnpmDeps_7_0 fetchPnpmDeps_6_0;
```
I think setting a default might cause breakages down the line if we ever need to drop an older version (like when it has a security vulnerability).

I have used 6_0 for now, as all packages use the 6.0 lock file format. We can probably use 7_0 too, but I thought this has the least likelihood of breakages

doronbehar replied:
> I think setting a default might cause breakages down the line if we ever need to drop an older version (like when it has a security vulnerability).
> 
> I have used 6_0 for now, as all packages use the 6.0 lock file format. We can probably use 7_0 too, but I thought this has the least likelihood of breakages

OK I see. It'd be nice perhaps to explain in a comment that we don't consider any lock file format as the "default" or "standard" or "most common" format, hence we don't set a default.

Scrumplex replied:
I have changed the function names again. Instead of reflecting the latest supported lockfile version, it's not just a number that will be incremented when needed. We can't use `pnpm 9.0.6` for projects that define `8.x.x` in their `package.json`

Scrumplex commented:
My last force push fixes a problem I caused in `nodePackages` where dependencies would be completely ignored.

doronbehar commented:
> My last force push fixes a problem I caused in `nodePackages` where dependencies would be completely ignored.
> 
> Edit: this introduced another problem. Perhaps I should run `nixpkgs-review` before pushing and not after ^^

Don't waste your memory on that - running nixpkgs-review on a PR that ofborg didn't fully evaluated yet requires a lot of memory, which will likely reduce your PC's performance. It also won't be much faster, so I'd stay patient instead :). You could of course at least check that the new attributes you introduced evaluate and build - that shouldn't require too much memory.

getchoo commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,101 @@
+    };
+  };
+
+  pnpms = builtins.mapAttrs (_: value: nodePackages.pnpm.overrideAttrs value) pinnedVersions;
```
```suggestion
  pnpmVersionFor = builtins.mapAttrs (_: nodePackages.pnpm.overrideAttrs) pinnedVersions;
```
i think the naming here would make more sense. we can also make an eta reduction (and maybe use `lib.const` if you want to be fancy)

getchoo commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,101 @@
+
+  pnpms = builtins.mapAttrs (_: value: nodePackages.pnpm.overrideAttrs value) pinnedVersions;
+
+  mkFetchPnpmDeps = version: let
+    pnpm = pnpms.${version};
```
```suggestion
  mkFetchPnpmDeps = lockfileVersion: let
    pnpm = pnpmVersionFor.${lockfileVersion};
```

getchoo commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,101 @@
+        fixupPhase = ''
+          runHook preFixup
+
+          rm -rf $out/v3/tmp
+          for f in $(find $out -name "*.json"); do
+            sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
+            jq --sort-keys . $f | sponge $f
+          done
```
```suggestion
          # Remove timestamp and sort the json files
          rm -rf $out/v3/tmp
          for f in $(find $out -name "*.json"); do
            sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
            jq --sort-keys . $f | sponge $f
          done
```
we should describe what this is doing. i used the comment from `pot`

getchoo commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,101 @@
+    pnpmConfigHook = makeSetupHook
+      {
+        name = "pnpm-${version}-config-hook";
+        propagatedBuildInputs = [pnpm];
```
```suggestion
        propagatedBuildInputs = [ pnpm ];
```

getchoo commented on pkgs/build-support/node/fetch-pnpm-deps/pnpm-config-hook.sh:
```diff
@@ -0,0 +1,44 @@
+    export HOME=$(mktemp -d)
+    export STORE_PATH=$(mktemp -d)
+
+    cp -Tr "$pnpmDeps" "$STORE_PATH"
+    chmod -R +w "$STORE_PATH"
```
before copying these over, we should probably make sure the lockfile in the source and dependency derivation are the same -- similar to [npmConfigHook](https://github.com/NixOS/nixpkgs/blob/c12b90be95acc0ab9102098d34ffb6dcb52cc55b/pkgs/build-support/node/build-npm-package/hooks/npm-config-hook.sh#L34)

Scrumplex replied:
We don't actually ship the `pnpm-lock.yaml` in the FOD. I don't think it matters much here, as a different lock file is certainly going to cause an installation failure, as not all packages will be available.

getchoo commented on pkgs/build-support/node/fetch-pnpm-deps/pnpm-config-hook.sh:
```diff
@@ -0,0 +1,44 @@
+    echo "Finished pnpmConfigHook"
+}
+
+postConfigureHooks+=(pnpmConfigHook)
```
`npmConfigHook` adds itself to `postPatchHooks`. is there a reason we're not following this example?

Scrumplex replied:
I personally felt like the configuration step is more fitting, but I am not sure tbh.

doronbehar replied:
> I personally felt like the configuration step is more fitting, but I am not sure tbh.

I think it can stay as it is.

Scrumplex commented:
Result of `nixpkgs-review pr 290715` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>geph.gui</li>
    <li>kiwitalk</li>
    <li>pot</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>youtube-music</li>
  </ul>
</details>

doronbehar commented:
@ofborg build youtube-music

Scrumplex committed:
vikunja: use pnpm.fetchDeps

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
Result of `nixpkgs-review pr 290715` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>geph.gui</li>
    <li>kiwitalk</li>
    <li>pot</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>youtube-music</li>
  </ul>
</details>

doronbehar commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,112 @@
+  # lockfile versions will require newer pnpm versions, we can will
+  # need to add more versions of this helper function in the future
+  pnpmVersions = {
+    "1" = nodePackages.pnpm.overrideAttrs {
+      name = "pnpm-8.15.5";
+      src = fetchurl {
+        url = "https://registry.npmjs.org/pnpm/-/pnpm-8.15.5.tgz";
+        hash = "sha512-sFGjLH5pWDO4SSbTspuMylclS1ifBknYmcbp0O22cLkex+KkNFm65zdZu1zmGcMmbxFr+THOItHvF1mn5Fqpbw==";
+      };
+    };
+    "2" = nodePackages.pnpm.overrideAttrs {
```
BTW we could also use a list instead:

```nix
pnpmVersions = [
  {
    lockfileVersion = "9";
    pnpmVersion = "9.0.6";
    srcHash = ...;
  }
  {
    lockfileVersion = "6";
    pnpmVersion = "8.15.5";
    srcHash = ...;
  }
];
```

Which is extremely readable.

doronbehar replied:
@Scrumplex This review comment of mine is the only comment I have left for this PR, otherwise it looks good.

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -10136,6 +10136,9 @@ with pkgs;
   importNpmLock = callPackages ../build-support/node/import-npm-lock { };
 
+  inherit (callPackages ../build-support/node/fetch-pnpm-deps { })
+    fetchPnpmDeps1 fetchPnpmDeps2;
```
I was thinking about it, and it'd be very nice if `pnpm` versions were available here - this would invite people to play with them, and they would find out these `pnpm` versions exist, only if they would come up in `nix search`. Perhaps another approach would be to define `pnpm8` and `pnpm9` attributes, and each will have a `pnpm${v}.fetchDeps` attribute.

Scrumplex replied:
While I agree with this, I think this would be a bit out of scope for this PR. Ideally we would promote pnpm to a standalone package (outside `nodePackages`) similar to `yarn` and provide some kind of `pnpmVersions` scope.

doronbehar reviewed (commented):
Some more ideas I was thinking upon.

doronbehar commented:
https://github.com/NixOS/GSoC/blob/989a0a84ec51df780998246030e9f06c1ed59e8b/ideas/2024.md

purepani commented:
It feels a bit weird to me to use the package installer itself to get deps, even if you pin the package manager. What if there's an security vulnerability in pnpm, for example, and it has to be removed? All the packages are now broken until it's fixed. Of the ecosystems I've dug around in, I can't immediately think of any that fetched dependencies using this technique as well. I'm sure people with more experience can fill in with more reasons, at least if what im saying is correct.

I was thinking of an architecture that would work and I was hoping to try when I had time: python has a fetchPypi function that we could emulate, so I think building of a fetchNpmRegsitry function could be a good idea. The hashes in pnpm, npm, and the old yarn lock give the same has as when you do `fetchurl`, so it wouldn't be tricky for any of those cases to utilize such a function.

In the case of the new yarn-berry, there is some argument in fetchurl that i'm forgetting the name of that lets you preprocess the output of fetchurl, which should let us rezip the yarn package in the same way that yarn does, so it should still be possible to have fetchurl calculate the same hash. This is the only case that's a bit trickier to implement as far as I can tell. 

If we have a fetchNpmRegistry function based on that idea, then basing the rest of the node ecosystem off of that function seems like a good strategy, at least to me.

purepani commented:
See https://github.com/NixOS/nixpkgs/pull/309740 for a basic implementation of a fetchNpmRegistry function.

doronbehar commented:
> Of the ecosystems I've dug around in, I can't immediately think of any that fetched dependencies using this technique as well.

Golang:

https://github.com/NixOS/nixpkgs/blob/6de580d56cb7fa273fa689a415a42fd48c897875/pkgs/build-support/go/module.nix#L125

> I was thinking of an architecture that would work and I was hoping to try when I had time: python has a fetchPypi function that we could emulate, so I think building of a fetchNpmRegsitry function could be a good idea.

The problem is not the matching hashes to URLs, but rather the fact that they are available only in `${src}/pnpm-lock.yaml`, meaning you have to evaluate and build `${src}` and only then you'd be able to read `pnpm-lock.yaml` - something that wouldn't be allowed on Hydra and ofborg, that disallow IFD.

This is a somewhat common problem in such ecosystems, and some package maintainers prefer vendoring the lock file into Nixpkgs, (e.g `Cargo.lock`, `package-lock.json`), which allows evaluating the hashes directly and not rely upon any ecosystem. Personally, I find that a big headache as a package maintainer, and I always prefer to avoid it, if possible. see also discussion [here](https://github.com/NixOS/nixpkgs/issues/296856#issuecomment-2003340070).

Having a `fetchPnpmRegistry` function would help if we are aiming towards vendoring `pnpm-lock.json` files for every package that needs it. Hence I personally prefer not putting efforts in that direction, because it is a maintenance headache to update these lock files for packages.

> What if there's an security vulnerability in pnpm, for example, and it has to be removed? All the packages are now broken until it's fixed.

I agree with that a bit. And we can write a program ourselves that will fetch all the sources for us, but that seems rather hard in the case of pnpm, because the format of the `pnpmDeps` derivations is very special. See for example the file layout of `pot.pnpm-deps` (from current `master` branch of Nixpkgs).

Even if you wouldn't mind vendoring `pnpm-lock.json` files into Nixpkgs, and you'd write Nix code that will evaluate these files and iterate all the dependencies and apply a `fetchPnpmRegistry` function to each, you'd have to eventually study the layout that pnpm uses and imitate it.

purepani commented:
You don't have to vendor the lock file necessarily, or do generation. 
`buildNpmPackage` doesn't do either of those things. The way it works is by fetching a lock file individually first. The only thing you have to do is say what the hash is.

doronbehar commented:
I'm not sure I understand what you are saying. I am looking at:

https://github.com/NixOS/nixpkgs/blob/6de580d56cb7fa273fa689a415a42fd48c897875/pkgs/build-support/node/fetch-npm-deps/default.nix#L150-L206

And I'm seeing that `prefetch-npm-deps` is accepting two arguments, `package-lock.json` and the output directory. I also inspected an example output of `prefetch-npm-deps` in `uivonim.npmDeps`, and I saw a very non-trivial file layout, that looks to me like it imitates how `npm` expects to find it's deps installed in it's cache directory. See:

https://github.com/NixOS/nixpkgs/blob/6de580d56cb7fa273fa689a415a42fd48c897875/pkgs/build-support/node/build-npm-package/hooks/npm-config-hook.sh#L75-L84

If I'll try to explain my previous comment in comparison to the npm ecosystem, it'll be summarized like this:
- The `buildNpmPackage` function isn't designed with a `fetchNpmRegistry` equivallent to `fetchPnpmRegistry`, but rather it uses a `prefetch-pnpm-deps` program written especially for NixOS.
- Using a `fetchPnpmRegistry` would require IFD due to evaluating the lock file to get all the hashes.
- Using a `fetchPnpmRegistry` (requires IFD) or writing a `prefetch-pnpm-deps` program (equivalent to `prefetch-npm-deps`) is pretty hard, and requires learning the layout of the cache directory that `pnpm install --frozen-lockfile --ignore-script` produces.

Scrumplex commented:
Even if we vendor `pnpm-lock.yaml`, it is a yaml file which we can't parse without converting it to json first. We either need IFD again, or we need to vendor the lock file in a Nix-friendly format. But I would argue that for the latter approach, it would be better to have some kind of `pnpn2nix` tool.



> You don't have to vendor the lock file necessarily, or do generation.
> `buildNpmPackage` doesn't do either of those things. The way it works is by fetching a lock file individually first. The only thing you have to do is say what the hash is.
> It fetches the lock file individually first, gets the dependencies, and then fetches the entire repo in a new derivation.

This is basically the approach of `fetchPnpmDeps` too, minus the fact that we use `pnpm` here instead of having a custom fetcher tool (see https://github.com/NixOS/nixpkgs/tree/master/pkgs/build-support/node/fetch-npm-deps)

szlend commented:
> What if there's an security vulnerability in pnpm, for example, and it has to be removed?

A project using a really old pnpm lockfile is very likely vulnerable itself. My understanding is you should be able to upgrade the fetcher without touching the lockfile with the caveat that the output hash could change. Here's [an example](https://github.com/pnpm/pnpm/issues/7934) of this not being the case and subsequently fixed upstream.

purepani commented:
I see that it's probably better to just get something working now; thanks for taking the time to help explain!

Scrumplex commented:
I feel like this is basically ready for review. The core functionality seems to be fairly stable and reliable. I am open to suggestions regarding the Nix code itself, as there are many ways to do this.

doronbehar commented:
I'm having trouble packaging https://github.com/bash-lsp/bash-language-server (that has [this `pnpm-lock.yaml`](https://github.com/bash-lsp/bash-language-server/blob/main/pnpm-lock.yaml) file). I'm getting:

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/4f0wfhml6zm3wq9150569riycdh4lc5r-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "installPhase" }
Running phase: installPhase
 WARN  using --force I sure hope you know what you are doing
 ERR_PNPM_LOCKFILE_CONFIG_MISMATCH  Cannot proceed with the frozen installation. The current "overrides" configuration doesn't match the value found in the lockfile

Update your lockfile using "pnpm install --no-frozen-lockfile"
```

Any idea what to do?

getchoo commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,115 @@
+      __functor = self: self.fetchPnpmDeps;
+    };
+
+  mkFetchPnpmDepsAttrs = fnVersion: pnpm: nameValuePair "fetchPnpmDeps${toString fnVersion}" (mkFetchPnpmDeps pnpm);
```
i would strongly suggest going back to the original naming scheme of lockfile version -- or alternative, use the pnpm version
using the really arbitrary index of a version in a list gives callers no real information on what that variant is for

purepani commented:
> I'm having trouble packaging https://github.com/bash-lsp/bash-language-server (that has [this `pnpm-lock.yaml`](https://github.com/bash-lsp/bash-language-server/blob/main/pnpm-lock.yaml) file). I'm getting:
> 
> ```
> @nix { "action": "setPhase", "phase": "unpackPhase" }
> Running phase: unpackPhase
> unpacking source archive /nix/store/4f0wfhml6zm3wq9150569riycdh4lc5r-source
> source root is source
> @nix { "action": "setPhase", "phase": "patchPhase" }
> Running phase: patchPhase
> @nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
> Running phase: updateAutotoolsGnuConfigScriptsPhase
> @nix { "action": "setPhase", "phase": "installPhase" }
> Running phase: installPhase
>  WARN  using --force I sure hope you know what you are doing
>  ERR_PNPM_LOCKFILE_CONFIG_MISMATCH  Cannot proceed with the frozen installation. The current "overrides" configuration doesn't match the value found in the lockfile
> 
> Update your lockfile using "pnpm install --no-frozen-lockfile"
> ```
> 
> Any idea what to do?

Which version of pnpm were you using?
It looks like this was a bug
https://github.com/pnpm/pnpm/issues/7184

doronbehar commented:
> Which version of pnpm were you using?
> It looks like this was a bug
> [pnpm/pnpm#7184](https://github.com/pnpm/pnpm/issues/7184)

Thanks for trying to help, I used simply `fetchPnpmDeps1` (which we don't like the name of it, but) it supposed to use pnpm `8.15.5`.

dotlambda commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,115 @@
+  # lockfile versions will require newer pnpm versions, we will need
+  # to add more versions of this helper function in the future
+  pnpmVersions = [
+    (nodePackages.pnpm.overrideAttrs {
```
Please don't use `nodePackages` but package pnpm separately.

NyCodeGHG commented:
This looks promising so far, I'm currently trying to repackage #290867 with fetchPnpmDeps, but it requires nodejs 18.
buildNpmPackage had an easy way to override the nodejs package, I don't think there is an easy way to do that without overriding nodePackages here, right?
Would be nice to have a similar option here as well.

doronbehar commented:
I'm also interested in checking [the issue I experienced](https://github.com/NixOS/nixpkgs/pull/290715#issuecomment-2101585758) with [bash-language-server](https://github.com/bash-lsp/bash-language-server), with nodejs 18.

Currently I see a few blockers for this PR:

- [ ] Better naming for the `fetchPnpmDeps*` attributes.
- [ ] Ability to override nodejs version used by pnpm.
- [ ] Package `pnpm` separately, not via an override to nodePackages ([this comment](https://github.com/NixOS/nixpkgs/pull/290715#discussion_r1597682901)).
- [ ] Verify that the new packages we want to use this fetcher for, work with it - bash-language-server, [webdriverio](https://github.com/NixOS/nixpkgs/pull/290867), and optionally: [surrealist](https://github.com/NixOS/nixpkgs/pull/307031).

dbaynard commented:
Not necessarily a blocker but this would be a nice-to-have: 

[`pnpm server --lock --store-dir "${pnpmDeps?}" start`](https://pnpm.io/cli/server) should enable subsequent pnpm calls (during development) to re-use the pnpm cache from the nix store. I'm not sure whether the cache created by this derivation _should be_ compatible with this command (different purposes) but it is worth being explicit that this is a non-goal.

doronbehar commented:
> [`pnpm server --lock --store-dir "${pnpmDeps?}" start`](https://pnpm.io/cli/server)

Are you talking about some kind of a development setup? Were you define in a `flake.nix` or a `shell.nix` a `pnpmDeps` using `fetchPnpmDeps`? I'm not sure I understand correctly.

dbaynard commented:
Yes — you've understood correctly. If the artifacts we create can be compatible with upstream tooling with no ongoing cost, it is worth thinking now.

One use case is development (I typically use nix only for language tools, not dependencies, but here the dependencies are provided as an optimization). Another is to run custom pnpm utilities in CI (though I'm sure there's a straightforward way to do so with the current code).

Scrumplex committed:
pnpm: init at 9.1.1

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
vesktop: use pnpm.fetchDeps

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
I have now added pnpm as a top-level package and moved `fetchPnpmDeps` to `pnpm.fetchDeps`. This means we can now pin pnpm (like I did with `pnpm_8`) and have versioned `fetchDeps` functions. (i.e. `pnpm_8.fetchDeps`)

nathanscully commented:
Thanks for working on this. I've been wanting to package pnpm projects for a while and this makes it much easier. I tested it out on a next's app inline here https://github.com/nathanscully/next-pnpm-nix and had no issues. I did have issues with it in a monorepo  using pnpm deploy here: https://github.com/nathanscully/pnpm-monorepo-test/tree/pnpm-nixpkgs for one of the packages, but the others which included workspace dependencies were ok. 

I'm getting the following error for the nextjs app in the mono repo which looks very similar to https://github.com/NixOS/nixpkgs/pull/309100 

```
       la/private/tmp/nix-build-frontend.drv-0/8b9z6rsmjic7cdchzkmiqdsksmq5l69d-source/apps/frontend:
       >  ERR_PNPM_NO_OFFLINE_META  Failed to resolve string-width@>=4.2.0 <5.0.0-0 in package mirror /private/tmp/nix-build-frontend.drv-0/tmp.fiUSfrvX7c/Library/Caches/pnpm/metadata/registry.npmjs.org/string-width.json
       >
       > This error happened while installing the dependencies of tailwindcss@3.4.3
       >  at sucrase@3.35.0
       >  at glob@10.3.10
       >  at jackspeak@2.3.6
       >  at @isaacs/cliui@8.0.2
       > Progress: resolved 13, reused 14, downloaded 0, added 0
       > /nix/store/g224xn9nwzf16jr94f1jx4jsnlazkl2a-stdenv-darwin/setup: line 140: pop_var_context: head of shell_variables not a function context
```

So I am guessing this is a pnpm specific issue related, just flagging it incase anyone else has any ideas to resolve it.

dotlambda commented on pkgs/development/tools/pnpm/default.nix:
```diff
@@ -0,0 +1,58 @@
+
+    meta = with lib; {
+      description = "Fast, reliable, and secure dependency management for javascript";
+      homepage = "https://classic.yarnpkg.com/";
```
?

Scrumplex replied:
totally forgot 😅

dotlambda commented on pkgs/development/tools/pnpm/default.nix:
Let's split this into `default.nix` and `generic.nix`.

dotlambda commented on pkgs/development/tools/pnpm/default.nix:
```diff
@@ -0,0 +1,58 @@
+, nodejs
+, stdenvNoCC
+, testers
+, withNode ? true
```
What's the reason for enabling/disabling this?

Scrumplex replied:
It was enabled in the Yarn top-level package so I thought I should do the same for pnpm.

SuperSandro2000 commented on pkgs/build-support/node/fetch-pnpm-deps/pnpm-config-hook.sh:
```diff
@@ -0,0 +1,44 @@
+    fi
+
+    if [ -z "${pnpmDeps-}" ]; then
+      echo "Error: 'pnpmDeps' should be set when using pnpmConfigHook."
```
```suggestion
      echo "Error: 'pnpmDeps' must be set when using pnpmConfigHook."
```

SuperSandro2000 commented on pkgs/build-support/node/fetch-pnpm-deps/pnpm-config-hook.sh:
```diff
@@ -0,0 +1,44 @@
+
+    echo "Installing dependencies"
+
+    if ! pnpm install --offline --frozen-lockfile --ignore-script; then
+      echo
+      echo "ERROR: pnpm failed to install dependencies"
+      echo
+
+      exit 1
+    fi
```
Shouldn't we fail anyway if the command fails and the error message doesn't seem to add much.

SuperSandro2000 commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,89 @@
+          "hash"
+          "pname"
+        ];
+        hash' =
+          if hash != "" then
+            { outputHash = hash; }
+          else
+            {
+              outputHash = "";
+              outputHashAlgo = "sha256";
+            };
```
What about just supporting hash?

Scrumplex replied:
I am not sure what you mean by this. I thought FODs need an `outputHash` attr

wegank replied:
It is just supporting `hash`, otherwise this would be `if hash != "" && sha256 != ""`.

SuperSandro2000 commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,89 @@
+            # Remove timestamp and sort the json files
+            rm -rf $out/v3/tmp
+            for f in $(find $out -name "*.json"); do
+              sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
```
Should we use subsituteInPlace --fail if nothing matches and abort?

Scrumplex replied:
I don't think every JSON file has this `checkedAt` value

wegank replied:
I don't think `[0-9]+` works in substituteInPlace either...

SuperSandro2000 commented on pkgs/by-name/ve/vesktop/package.nix:
```diff
@@ -33,66 +30,18 @@ stdenv.mkDerivation (finalAttrs: {
-      outputHashMode = "recursive";
-      outputHash = "sha256-PogE8uf3W5cKSCqFHMz7FOvT7ONUP4FiFWGBgtk3UC8=";
-    };
+  pnpmDeps = pnpm_8.fetchDeps {
```
Maybe we should uplift this to a wrapper like buildNpmPackage to avoid having to inherit so many variables here

doronbehar replied:
> Maybe we should uplift this to a wrapper like buildNpmPackage to avoid having to inherit so many variables here

That would be ideal, but I think this PR has a lot of value in its current form without such a function defined. In the meantime, perhaps the lack of an equivalent to `buildNpmPackage` for pnpnm should be noted in the docs, which I hope we'll write once this PR stabilizes.

SuperSandro2000 commented on pkgs/by-name/ve/vesktop/package.nix:
```diff
@@ -33,66 +30,18 @@ stdenv.mkDerivation (finalAttrs: {
-      outputHash = "sha256-PogE8uf3W5cKSCqFHMz7FOvT7ONUP4FiFWGBgtk3UC8=";
-    };
+  pnpmDeps = pnpm_8.fetchDeps {
+    inherit (finalAttrs) pname version src patches ELECTRON_SKIP_BINARY_DOWNLOAD;
```
Maybe set ELECTRON_SKIP_BINARY_DOWNLOAD as default? otherwise it might be forgotten often.

Scrumplex replied:
Seems like this doesn't have any effect for `pnpm.fetchDeps` so I just removed this.

SuperSandro2000 commented on pkgs/by-name/vi/vikunja/package.nix:
```diff
@@ -121,8 +69,10 @@ buildGoModule {
   vendorHash = "sha256-d4AeQEAtPqMDe5a5aKhCe3i3pDXAMZJkJXxfcAFTx7A=";
 
+  inherit frontend;
+
   prePatch = ''
-    cp -r ${frontend} frontend/dist
+    cp -r $frontend frontend/dist
```
Relying on bashrc here is not that nice. We should probably use finalAttrs if overrides are important.

Scrumplex replied:
This is in a `buildGoModule` derivation without support for function based attrs :(

SuperSandro2000 commented on pkgs/applications/networking/geph/default.nix:
```diff
@@ -60,42 +61,12 @@ in
     gephgui-wry = rustPlatform.buildRustPackage {
       pname = "gephgui-wry";
-      inherit version src;
+      inherit version;
+      inherit (finalAttrs) src;
```
We should take both from finalAttrs

SuperSandro2000 commented on pkgs/applications/networking/geph/default.nix:
```diff
@@ -105,10 +76,18 @@ in
+      pnpmDeps = pnpm_8.fetchDeps {
+        inherit (finalAttrs) pname version src;
+        sourceRoot = "${finalAttrs.src.name}/gephgui-wry/gephgui";
+
```
```suggestion
```

doronbehar reviewed (commented):
Looking much better :) I'm still experiencing [the error I quoted with `bash-language-server`](https://github.com/NixOS/nixpkgs/pull/290715#issuecomment-2101585758) though :/.

doronbehar commented on pkgs/development/tools/pnpm/default.nix:
```diff
@@ -1,7 +1,8 @@
 , fetchurl
+, mkFetchPnpmDeps
 , nodejs
-, stdenvNoCC
```
> Unnecessary change in the `fetchPnpmDeps: init` commit.

Was commenting an outdated version of the PR.

pluiedev commented on pkgs/build-support/node/fetch-pnpm-deps/default.nix:
```diff
@@ -0,0 +1,89 @@
+            # Remove timestamp and sort the json files
+            rm -rf $out/v3/tmp
+            for f in $(find $out -name "*.json"); do
+              sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
+              jq --sort-keys . $f | sponge $f
```
Couldn't we implement this in `jq` itself?
```suggestion
              jq --sort-keys "del(.. | .checkedAt?)" $f | sponge $f
```

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -10080,6 +10080,9 @@ with pkgs;
   importNpmLock = callPackages ../build-support/node/import-npm-lock { };
 
+  inherit (callPackages ../build-support/node/fetch-pnpm-deps { })
+    mkFetchPnpmDeps;
```
> This seems a bit weird.. Why not `fetchPnpmDeps = pnpm.fetchDeps`? and/or `pnpmConfigureHook = pnpm.configHook`?

This still bothers me BTW...

doronbehar replied:
> I don't think we should have a top-level function for this. It is very possible that future iterations of this function _will_ break existing packages, so we shouldn't encourage people to use a top-level function that might be outdated after some time.

OK I agree with that, but I don't understand why would someone want to use `mkFetchPnpmDeps`, and not `pnpm_{8,9}.fetchDeps`?

doronbehar replied:
> mkFetchPnpmDeps is probably not very useful in the top-level, yes. I wasn't sure where to put it, because it seems like build helpers usually go into `pkgs/build-support`, but I didn't think `pnpm` itself should be in there?
> 
> So either I move the `fetchDeps` function into `pkgs/development/tools/pnpm` or I move `pnpm` alongside `fetchDeps` to `pkgs/build-support/node/pnpm`

Oh I think I understand, you import `mkFetchPnpmDeps` in `pnpm/generic.nix`... I think the simplest thing would be to not even define `mkFetchPnpmDeps`, and put the contents of this function in the `passthru` of `pnpm/generic.nix`.

doronbehar replied:
You could also define the `mkFetchPnpmDeps` function in a separate file near `pnpm/generic.nix` IMO.

doronbehar replied:
I still oposing exposing `mkFetchPnpmDeps`.

Scrumplex commented:
I have added some brief documentation. If you think something needs documentation, feel free to let me know

doronbehar commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -310,6 +310,64 @@ See `node2nix` [docs](https://github.com/svanderburg/node2nix) for more info.
+
+Pnpm is available as the top-level package `pnpm`. Additionally, there are variants pinned to certain major versions, like `pnpm_8` and `pnpm_9`.
+
+When packaging an application that includes a `pnpm-lock.yaml`, you need to fetch the pnpm store for that project using a fixed-output-derivation. The `pnpm` derivations come with a `fetchDeps` function and a `configHook` setup hook respectively.
+
+A package containing a `package.json` and a `pnpm-lock.yaml` can be packaged like this:
```
Minor rephrasing:

```suggestion
When packaging an application that includes a `pnpm-lock.yaml`, you need to fetch the pnpm store for that project using a fixed-output-derivation. The functions `pnpm_8.fetchDeps` and `pnpm_9.fetchDeps` can create this pnpm store derivation. In conjunction, the setup hooks `pnpm_8.configHook` and `pnpm_9.configHook` will tell `pnpm` to use the above prefetched dependencies store. Here is an example for a package that contains a `package.json` and a `pnpm-lock.yaml` files using the above `pnpm_` attributes:
```

doronbehar commented on doc/languages-frameworks/javascript.section.md:
````diff
@@ -310,6 +310,64 @@ See `node2nix` [docs](https://github.com/svanderburg/node2nix) for more info.
+A package containing a `package.json` and a `pnpm-lock.yaml` can be packaged like this:
+
+```nix
+# pkgs.callPackage ... { pnpm = pnpm_9; }
````
What does this mean?

doronbehar commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -310,6 +310,64 @@ See `node2nix` [docs](https://github.com/svanderburg/node2nix) for more info.
+
+  src = ...;
+
+  nativeBuildInputs = [nodejs pnpm.configHook];
```
```suggestion
    nativeBuildInputs = [
      nodejs
      pnpm_8.configHook
    ];
```

doronbehar commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -310,6 +310,64 @@ See `node2nix` [docs](https://github.com/svanderburg/node2nix) for more info.
 
+### pnpm {#javascript-pnpm}
+
+Pnpm is available as the top-level package `pnpm`. Additionally, there are variants pinned to certain major versions, like `pnpm_8` and `pnpm_9`.
```
We should say a word here about the lock file version link to pnpm versions.

doronbehar commented on doc/languages-frameworks/javascript.section.md:
````diff
@@ -310,6 +310,64 @@ See `node2nix` [docs](https://github.com/svanderburg/node2nix) for more info.
+})
+```
+
+NOTE: It is highly recommended to use a pinned version of pnpm (i.e. `pnpm_8` or `pnpm_9`), to increase future reproducibility.
````
Maybe mention the lockfile version link to pnpm versions here as well, because it is an important detail.

doronbehar commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -310,6 +310,64 @@ See `node2nix` [docs](https://github.com/svanderburg/node2nix) for more info.
+
+NOTE: It is highly recommended to use a pinned version of pnpm (i.e. `pnpm_8` or `pnpm_9`), to increase future reproducibility.
+
+In case you are patching `package.json` or `pnpm-lock.yaml`, make sure to pass `finalAttrs.patches` to the function as well.
```
Maybe this will be a bit clearer:

```suggestion
In case you are patching `package.json` or `pnpm-lock.yaml`, make sure to `inherit (finalAttrs) patches` when calling `fetchDeps`.
```

doronbehar reviewed (commented):
Great docs! Well done @Scrumplex on all the work. One additional small note: the 2nd commit message should be `pnpm.fetchDeps: init` and not `fetchPnpmDeps: init`.

doronbehar commented:
BTW I had a little bit of progress with `bash-language-server` - the problem was that I pulled an old git tag of it (haven't checked what `nix-init` was defaulting to). I pushed my progress into the last commit of this branch, which is based upon the branch of this PR:

https://github.com/doronbehar/nixpkgs/commits/pkg/bash-language-server--pnpm.fetchDeps

Still though, I ran into weird errors with all kinds of dependencies like `husky`, `cross-env` and `rollup` not found, and I don't understand why `pnpm.fetchDeps` didn't pull them. I sort of managed to make `npmInstallHook` pretend these executables exist, and eventually the build finished, but it produced an executable that fails with:

```
node:internal/modules/cjs/loader:1146
  throw err;
  ^

Error: Cannot find module 'vscode-languageserver/node'
Require stack:
- /nix/store/91pbbmjkjn2jqmbfd2yqky84fcan2g35-bash-language-server-5.3.3/lib/node_modules/bash-language-server/out/cli.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1143:15)
    at Module._load (node:internal/modules/cjs/loader:984:27)
    at Module.require (node:internal/modules/cjs/loader:1231:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/nix/store/91pbbmjkjn2jqmbfd2yqky84fcan2g35-bash-language-server-5.3.3/lib/node_modules/bash-language-server/out/cli.js:15:13)
    at Module._compile (node:internal/modules/cjs/loader:1369:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1427:10)
    at Module.load (node:internal/modules/cjs/loader:1206:32)
    at Module._load (node:internal/modules/cjs/loader:1022:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/nix/store/91pbbmjkjn2jqmbfd2yqky84fcan2g35-bash-language-server-5.3.3/lib/node_modules/bash-language-server/out/cli.js'
  ]
}

Node.js v20.12.2
```

And it is also peculiar because the project does include this `vscode-languageserver/node` dependency.

nyabinary commented:
Anything stopping this from being merged apart from the comments above?

Atemu commented:
What about the TODOs @Scrumplex?

Scrumplex commented:
Time to continue working on this! :D

Scrumplex commented:
I have moved the `fetchDeps` and `configHook` out of build support now.

Scrumplex committed:
pnpm.fetchDeps: add serve script

This script allows users to reuse the cached dependencies outside of
derivations.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
I have added a serve script! Now you can reuse dependencies from the prefetched store by running `pnpm run .#<package>.pnpmDeps.serve`

purepani commented:
> I have added a serve script! Now you can reuse dependencies from the prefetched store by running `pnpm run .#<package>.pnpmDeps.serve`

I assume you mean `nix run` not `pnpm run` 😁

Scrumplex commented:
> I assume you mean `nix run` not `pnpm run` 😁

Oops! Yep :D

doronbehar commented:
> I have moved the `fetchDeps` and `configHook` out of build support now.

What do you mean by "out of build support"?

Scrumplex commented:
> > I have moved the `fetchDeps` and `configHook` out of build support now.
> 
> What do you mean by "out of build support"?

I have moved it to `pkgs/development/tools/pnpm/fetch-deps` now :D

doronbehar commented:
Thanks @Scrumplex for you dedication :) I went over all the comments to the PR and everything looks good, besides [the issue I experienced with `bash-language-server`](https://github.com/NixOS/nixpkgs/pull/290715#issuecomment-2132073016) :disappointed: . I also tried this:

```nix
stdenv.mkDerivation (finalAttrs: {
  pname = "bash-language-server";
  version = "5.3.3";

  src = fetchFromGitHub {
    owner = "bash-lsp";
    repo = "bash-language-server";
    rev = "server-${finalAttrs.version}";
    hash = "sha256-4+imeVjosEspFhWzjtUp2IZx7aZIx56g8M03V3dEGVs=";
  };

  pnpmDeps = pnpm_9.fetchDeps {
    inherit (finalAttrs) pname version src;
    hash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
    sourceRoot = "${finalAttrs.src.name}/server";
  };
```

And it failed with:

```
 EACCES  EACCES: permission denied, mkdir '/build/source/node_modules'
```

However, using:

```nix
  pnpmDeps = pnpm_9.fetchDeps {
    inherit (finalAttrs) pname version src;
    hash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
    preInstall = ''
      cd server
    '';
  };
```

Didn't fail, and produced the same hash, and then the same issues as before. What do you think about that @Scrumplex ? Here's the source code for bash-language-server: https://github.com/bash-lsp/bash-language-server .

honnip commented:
I believe that monorepo is not supported for now.

Scrumplex commented:
Yeah, we need to change how this works for monorepos/workspaces.

Normal projects:
- package.json, pnpm-lock.yaml and node_modules are in the same directory

Changing `sourceRoot` of `fetchDeps` is sufficient, in case the pnpm project isn't in the root of `src`.

Workspace projects:
- pnpm-lock.yaml and node_modules are in workspace root
- package.json is in a subdirectory

Changing `sourceRoot` to a subdirectory leads to a read only project root and therefore stops pnpm from creating a `node_modules` directory.
I tried changing the `node_modules` path using `--modules-dir` but it seemed like it didn't do anything?

doronbehar commented:
OK so since there are many people who want this PR to get in, and it is pretty much ready IMO, monorepos/workspaces support shouldn't be a blocker. It'd be OK with me that for the meantime, we'd officially state in the docs that this is not supported.

We can work on it some other time, unless @Scrumplex you'd like to keep giving more effort on this.

Scrumplex commented:
It would be nice to get workspaces working, but I don't think there is an optimal way to do it, so it would require some discussion.

I would love to get some more reviews, though! It's always easier to change things in a PR than in `master`, so I'd like to have some feedback for my decisions, like the versioned packages.

doronbehar commented:
> It would be nice to get workspaces working, but I don't think there is an optimal way to do it, so it would require some discussion.

OK. Then let's put a sentence about this in the docs, and wait for others to give their final reviews.

> I would love to get some more reviews, though!

A review from @dbaynard about the experience of the `nix run .#<package>.pnpmDeps.serve` interface would be nice.

> It's always easier to change things in a PR than in `master`, so I'd like to have some feedback for my decisions, like the versioned packages.

I agree. If I remember correctly, @pluiedev, @uncenter, @gepbird, @wegank and @getchoo have expressed their opinion before about the previous attribute names, so it'd be nice to get a :+1: from theme on the current naming of `pnpm_{8,9}`.

linsui commented:
There are binaries in the pnpm tgz. Is it `binaryNativeCode`?

doronbehar commented:
> There are binaries in the pnpm tgz.

Just out of curiosity, how did you find out this?

doronbehar committed:
pnpm_{8,9}: remove uneeded binary files from src

doronbehar committed:
doc/javascript: pnpm: mention lack of monorepos/workspaces support

doronbehar commented:
> > Just out of curiosity, how did you find out this?
> 
> Open the src tgz in file roller... There are some .node files in `package/dist/`.

Hmm they seem to be targetting only Darwin and Windows... I managed to remove them and `pnpm_{8,9}` built fine for me. Let's see what CI says for Darwin:

@ofborg build pnpm_8 pnpm_9

Also, I pushed a small addition to the docs mentioning the lack of support for monorepos and worksapces, so I think this will be ready once CI is greener.

linsui commented:
I'm not sure how to give :+1: but your commit is good to me. :) Thanks!

doronbehar commented:
> I'm not sure how to give 👍 but your commit is good to me. :) Thanks!

Such a comment is exactly what I meant :), thanks.

honnip commented:
I can confirm that kiwitalk is working fine. Thanks.

But `bash-language-server` doesn't seem to work.
```shell
->> bash-language-server 
node:internal/modules/cjs/loader:1146
  throw err;
  ^

Error: Cannot find module 'vscode-languageserver/node'
Require stack:
- /nix/store/0qihgy16wg7qh01fs8rswb0h230kfdn2-bash-language-server-5.3.3/lib/node_modules/bash-language-server/out/cli.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1143:15)
    at Module._load (node:internal/modules/cjs/loader:984:27)
    at Module.require (node:internal/modules/cjs/loader:1231:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/nix/store/0qihgy16wg7qh01fs8rswb0h230kfdn2-bash-language-server-5.3.3/lib/node_modules/bash-language-server/out/cli.js:15:13)
    at Module._compile (node:internal/modules/cjs/loader:1369:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1427:10)
    at Module.load (node:internal/modules/cjs/loader:1206:32)
    at Module._load (node:internal/modules/cjs/loader:1022:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/nix/store/0qihgy16wg7qh01fs8rswb0h230kfdn2-bash-language-server-5.3.3/lib/node_modules/bash-language-server/out/cli.js'
  ]
}

Node.js v20.12.2
```

doronbehar commented:
> I can confirm that kiwitalk is working fine. Thanks.
> 
> But `bash-language-server` doesn't seem to work.

Aish I didn't mean to push that, thanks for noticing it quickly and for reviewing.

doronbehar commented:
It seems no body has anything further to add / review. Merging :tada:.

SuperSandro2000 commented:
We should have put the overrides for pnpm for the consuming packages inside the packages. pnpm is really bitchy about a wrong version and that needs to be closely tracked.

doronbehar commented:
> We should have put the overrides for pnpm for the consuming packages inside the packages.

I don't understand this sentence...

SuperSandro2000 commented:
They are right now in all-packages where you can easily miss them and wonder why your pnpmDeps build is failing for no apparent reason, since the error message is so bad.

see #317739 for my idea on how to improve that

NyCodeGHG commented:
I'm trying to package renovate using pnpm.fetchDeps, but I get the error `ERR_PNPM_NO_OFFLINE_META`  when executing `pnpm deploy`, looks like this is the same problem as in #309100

I can only reproduce it inside the nix sandbox, outside of it (and in a network namespace without internet access), it works fine, not sure how to debug this further.

build log:
https://gist.github.com/NyCodeGHG/073f40768d21796fac69cd98b8e0a8df

GetPsyched commented:
Dropping in here to ask about monorepo/workspaces support. I'm packaging a monorepo project and noticed from the discussion here that it isn't supported yet.

Is there an issue or, better yet, a draft PR that I can track for this?

pluiedev commented:
> Dropping in here to ask about monorepo/workspaces support. I'm packaging a monorepo project and noticed from the discussion here that it isn't supported yet.
> 
> Is there an issue or, better yet, a draft PR that I can track for this?

I believe you're looking for #316908

doronbehar commented:
> I'm trying to package renovate using pnpm.fetchDeps, but I get the error `ERR_PNPM_NO_OFFLINE_META` when executing `pnpm deploy`, looks like this is the same problem as in #309100

@NyCodeGHG since the issue #309100 was diagnosed as a workspaces issue, perhaps your issue with `renovate` is the same? See:

- https://github.com/NixOS/nixpkgs/issues/316908

NyCodeGHG commented:
I'm not sure if thats the problem, renovate [has a pnpm-workspaces.yaml file without a single member](https://github.com/renovatebot/renovate/blob/7973dbfc305229dc449aa05d4ecb10d5b15cc18a/pnpm-workspace.yaml#L1)
don't ask me why

doronbehar commented:
> I'm not sure if thats the problem, renovate [has a pnpm-workspace.yaml file without a single member](https://github.com/renovatebot/renovate/blob/7973dbfc305229dc449aa05d4ecb10d5b15cc18a/pnpm-workspace.yaml#L1) don't ask me why

Does it help to simply remove it? :)

NyCodeGHG commented:
i just tried, no.
looks like they have that file to prevent pnpm from also installing dependencies from test fixtures

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/help-wanted-planning-for-summer-of-nix-2025/47556/1

Artturin commented:
This has to be applied to node2nix otherwise it wont stick

https://github.com/orgs/nix-community/discussions/1523
